# Ep 1 
Aluno : Daniel Carlos Souza Santos <br />
$N_{usp}$ : 13686330

### Dependências 
Kotlin.math é uma biblioteca que contém os métodos matemáticos exigidos para o ep. Letsplot é uma biblioteca para exportação gráfica

In [34]:
%use lets-plot
import kotlin.math.pow
import kotlin.math.cos
import kotlin.math.sin
import kotlin.math.sign
import org.jetbrains.letsPlot.export.ggsave
import org.jetbrains.letsPlot.geom.geomLine
import org.jetbrains.letsPlot.letsPlot
import org.jetbrains.letsPlot.label.ggtitle
import org.jetbrains.letsPlot.label.xlab
import org.jetbrains.letsPlot.label.ylab
import org.jetbrains.letsPlot.scale.scaleXContinuous


## Exercício 1 
### Ítem (a)
Para resolver a equação $ x^{\frac{3}{4}} - cos(x^{2}) = 0 $ vamos dizer que $ x_1 $ é um ponto mais a esquerda que $ x_2 $ sempre. Seja $ f(x) = x^{\frac{3}{4}} - cos(x^{2}) $. O programa abaixo cria uma função que calcula a média de 2 valores e outra que corresponde a função do enunciado. Plotando a função do enunciado no seu respectivo domínio.

In [35]:
fun target( x : Double ) : Double {
    return x.pow(3.0/4.0) - cos(x.pow(2.0))
}
fun x_med( a : Double, b : Double) : Double {
    return ( a + b )/ 2.0
}

val xs = (0..300).map { it / 10.0 }
var ys = xs.map { target(it) }

val data = mapOf(
    "x" to xs,
    "y" to ys
)

val plot = letsPlot(data) + geomLine {
    x = "x"
    y = "y"
} + ggtitle("Plot da função target(x)") + xlab("x") + ylab("target(x)")

plot



In [36]:
var x_1 : Double = 0.0
var x_2 : Double = 1.0
val err  = 0.000001
// Essa parte do código inicializa x1 e x2. Após atualiza x1 e xx2 conforme indicado pelo método da bissecção
while (Math.abs(x_1 - x_2) > err){
    if(sign(target(x_1)) ==  sign(target(x_med(x_1, x_2))) ){
        x_1 = x_med(x_1, x_2)
    }else{
        x_2 = x_med(x_1, x_2)
    }
    println("x1: $x_1")
    println("x2: $x_2")
    println("xMed: ${x_med(x_1, x_2)}")
}


x1: 0.5
x2: 1.0
xMed: 0.75
x1: 0.75
x2: 1.0
xMed: 0.875
x1: 0.75
x2: 0.875
xMed: 0.8125
x1: 0.75
x2: 0.8125
xMed: 0.78125
x1: 0.75
x2: 0.78125
xMed: 0.765625
x1: 0.765625
x2: 0.78125
xMed: 0.7734375
x1: 0.7734375
x2: 0.78125
xMed: 0.77734375
x1: 0.7734375
x2: 0.77734375
xMed: 0.775390625
x1: 0.7734375
x2: 0.775390625
xMed: 0.7744140625
x1: 0.7734375
x2: 0.7744140625
xMed: 0.77392578125
x1: 0.77392578125
x2: 0.7744140625
xMed: 0.774169921875
x1: 0.774169921875
x2: 0.7744140625
xMed: 0.7742919921875
x1: 0.7742919921875
x2: 0.7744140625
xMed: 0.77435302734375
x1: 0.77435302734375
x2: 0.7744140625
xMed: 0.774383544921875
x1: 0.774383544921875
x2: 0.7744140625
xMed: 0.7743988037109375
x1: 0.774383544921875
x2: 0.7743988037109375
xMed: 0.7743911743164062
x1: 0.7743911743164062
x2: 0.7743988037109375
xMed: 0.7743949890136719
x1: 0.7743949890136719
x2: 0.7743988037109375
xMed: 0.7743968963623047
x1: 0.7743949890136719
x2: 0.7743968963623047
xMed: 0.7743959426879883
x1: 0.7743959426879883
x2: 0

Não existem outras raizes pois a função não é devinida para numéros negativos e é crescente na medida que x cresce. <br />
Resposta encontrada : 0.7743964195251465


## Item (b)
A derivada da função do ítem (a) é : $f'(x) = \frac{3}{4}x^{\frac{-1}{4}} + 2x*sen(x^{2})$, a qual foi calculada analíticamente. Para o método de Newton vamos usar o programa a seguir cujo o algoritmo replica exatamente $x_{n+1} = x_n - \frac{f(x_n)}{f'(x_n)}$ parando apenas quando $f(x_{n+1})$ está convenientemente próxima de 0. Ao final pode-se notar que ambos os métodos convergiram para valores próximos.


In [37]:
fun target_line(x :Double) :Double {
    return (3.0/4.0)*x.pow(-1.0/4.0) + 2*x*sin(x.pow(2.0))
}
var x_n : Double = 1.0 // tem que ser diferente de 0
var x_j : Double = 0.0 // x_n+1

while(Math.abs(target(x_n)) > err){
    if(x_n > err){
        x_j = x_n - target(x_n)/target_line(x_n)
        println("xn = ${x_n} e x_n+1 = ${x_j} \n")
        x_n = x_j
        
    }
    else{
        break
    }
}

xn = 1.0 e x_n+1 = 0.8110527501794648 

xn = 0.8110527501794648 e x_n+1 = 0.7755249564332772 

xn = 0.7755249564332772 e x_n+1 = 0.7743977471605691 

xn = 0.7743977471605691 e x_n+1 = 0.7743966632733174 




Resposta encontrada : 0.7743966632733174 

## Item (c) 
Como $V(r) = \frac{-e^2}{4\pi \epsilon_0 r} + V_0e^{\frac{-r}{r_0}}$ e também $F(r) = \frac{-e^2}{4\pi \epsilon_0 r^2} + \frac{V_0}{r_o}e^{\frac{-r}{r_0} $ os gráficos do potêncial e da força, pedidos no ítem $I$ ficam :  

In [38]:
fun potencial( x : Double) :Double { 
    return ( -14.4/x ) + 1380* exp(-x/0.328)  
}
fun force( x : Double) :Double{
    return ( -14.4/x.pow(2.0) ) + (1380/0.328)* exp(-x/0.328)
}
var ys2 = xs.map {force(it)}
ys = xs.map { potencial(it) }

val data = mapOf(
    "x" to xs,
    "y" to ys,
    "y2" to ys2
    
)

val plot = letsPlot(data) + geomLine(color = "dark_green", alpha = .7) {
    x = "x"
    y = "y"
} + geomLine(color = "red", alpha = .7){
    x = "x"
    y = "y2"
}+ggtitle("Plot das funções Potencial(x) e Force(x), X em Angstrong e Potencial em eV") + ggsize(700, 500) +xlab("x") + ylab("potencial(x)")+ scaleFillBrewer(type = "seq") + theme(panelGridMajorX = "blank")+ scaleXContinuous()
plot

// Distancias em Angstrong e potenciais em eletron-volts

A função Potencial, em verde, foi plotada ao lado da função Força em vermelho. 
<p color = 'green'> Potencial </p> <p color = 'red'> Força </p><br />
Abaixo vem uma implementação literal do método das secantes. Ele visa resolver a parte $II$ do ítem c.


In [39]:
// Parte 2 do ítem (c). Para achar a região em que F(r) = 0 pelo método das secantes façamos : 
var x_a : Double = 1.0 // x_n+1
var x_b : Double = 2.0 // x_n
var x_c : Double = 3.0 // x_n-1
// err é um número arbitrariamente perto de 0, o qual nesse caso servirá para não fazer a segurança do programa.
while (Math.abs(force(x_a)) > err ){
    if(x_a > err && x_b > err && x_c > err ){
        x_a = x_b - (force(x_b)*(x_b - x_c))/(force(x_b) - force(x_c))
        x_c = x_b
        x_b = x_a
        println("x_a = ${x_a}, x_b = ${x_b}, x_c = ${x_c}")
    }else{
        break
    }
} // as unidades aqui usadas foram as mesmas indicadas no programa. (ev e Å)

x_a = 2.8357626582048745, x_b = 2.8357626582048745, x_c = 2.0
x_a = 2.7086896209157096, x_b = 2.7086896209157096, x_c = 2.8357626582048745
x_a = 2.086616863631506, x_b = 2.086616863631506, x_c = 2.7086896209157096
x_a = 2.5963071775342623, x_b = 2.5963071775342623, x_c = 2.086616863631506
x_a = 2.5291460141555033, x_b = 2.5291460141555033, x_c = 2.5963071775342623
x_a = 2.4239567555493937, x_b = 2.4239567555493937, x_c = 2.5291460141555033
x_a = 2.453990502824868, x_b = 2.453990502824868, x_c = 2.4239567555493937
x_a = 2.4501966676359572, x_b = 2.4501966676359572, x_c = 2.453990502824868
x_a = 2.450005803357345, x_b = 2.450005803357345, x_c = 2.4501966676359572
x_a = 2.4500072105655226, x_b = 2.4500072105655226, x_c = 2.450005803357345


Resposta encontrada : 2.4500072105655226 Å